WordPress/src/main/java/org/wordpress/android/util/experiments/ExPlat.kt

```diff
@@ -0,0 +1,52 @@
+package org.wordpress.android.util.experiments
+
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+import org.wordpress.android.fluxc.model.experiments.Assignments
+import org.wordpress.android.fluxc.store.ExperimentStore
+import org.wordpress.android.fluxc.store.ExperimentStore.FetchAssignmentsPayload
+import org.wordpress.android.fluxc.store.ExperimentStore.Platform
+import org.wordpress.android.fluxc.utils.AppLogWrapper
+import org.wordpress.android.modules.APPLICATION_SCOPE
+import org.wordpress.android.util.AppLog.T
+import javax.inject.Inject
+import javax.inject.Named
+import javax.inject.Singleton
+
+@Singleton
+class ExPlat
+@Inject constructor(
+    private val experimentStore: ExperimentStore,
+    private val appLog: AppLogWrapper,
+    @Named(APPLICATION_SCOPE) private val coroutineScope: CoroutineScope
+) {
+    private val platform = Platform.WORDPRESS_COM
+
+    fun refresh() {
+        getAssignments(shouldFetchIfStale = true)
```

Comment:
how do we make sure that the experiment is not fetched in the middle of a user process and doesn't change the behaviour of the current flow? For example you open an activity with a control, the app goes to the background, goes back, experiments are refreshed, you get assigned to a variant A, you click on a button and even though your View is "control", the action could be "variant A". We need a pull based cache for the experiments - once the value is read, it should never be changed as long as the app is alive.

Reply:
If I'm not missing something, I believe this would depend on the experiment. Sometimes we will want to update the UI as soon as possible, some other times we won't.

In the example you gave, we would basically have two options. The first one would be to re-render the view when the app came back from background using the "variant A". If we don't want to do that because, let's say, the user was in the middle of a flow where updating the UI would be disrupting, then we should keeping using the control variant for the duration of that flow, but this wouldn't be handled by ExPlat.

We would also have to make sure that the exposure event for that experiment was triggered as soon as the user entered the flow and that it wasn't triggered again when the app came back from the background. As the variant changed, we can assume that the correct variant was only assigned to the user _after_ he had already entered the flow. This means that all actions the user takes during that flow will be discarded by ExPlat and won't count towards the experiment analysis.

Reply:
I think this is error prone. This requires us to keep the current experiment state somewhere out of the experiment. For example an experiment in the login flow would require us to store the value when the flow starts somewhere outside of the `Experiment` class. 

> We would also have to make sure that the exposure event for that experiment was triggered as soon as the user entered the flow and that it wasn't triggered again when the app came back from the background. As the variant changed, we can assume that the correct variant was only assigned to the user after he had already entered the flow. This means that all actions the user takes during that flow will be discarded by ExPlat and won't count towards the experiment analysis.

Is the exposure event triggered manually? Or does it happen when the value is fetched? If we implement a pull-based-cache, solving this would be easy. Whenever we check the experiment value and it's not set yet, we can send the exposure event.

Reply:
>I think this is error prone.

At this point I believe most experiments that escape the most common cases are probably error prone in one way or another. This is by design. That said, I think I understand the point you're making and I agree with it.

>This requires us to keep the current experiment state somewhere out of the experiment.

By default, yes, that's true. But we can always implement an experiment that manages its own state later on, as needed.

This would probably require some changes to the way we're currently refreshing assignments, but that's not a problem. In fact, one the reasons this implementation is so lax at this point, is because we're supposed to change it as we learn what we actually need for each type of experiment. If we can anticipate some of those needs, great, if not, that's ok too.

Once we're in a place where we feel the implementation is solid enough for most of the cases we're trying to work with, then we would move this to its own library, so it can be used by other apps as well. Until then, all of this should be considered experimental.

>Is the exposure event triggered manually? Or does it happen when the value is fetched? 

It is triggered manually. Keep in mind that this event could be any of the existing Tracks events we already send. It is not necessarily an event created specifically for the experiment. In this experiment, we're using the `reader_accessed` event, for example.

>If we implement a pull-based-cache, solving this would be easy. Whenever we check the experiment value and it's not set yet, we can send the exposure event.

Actually, the exposure event should be sent whenever the user is exposed to the experiment, regardless of the variation they are seeing. But again, I understand this is besides the point you're making.